test/kennel/filter_test.rb

```diff
@@ -4,61 +4,90 @@
-  end
 
-  context "with project, with tracking_id" do
-    context "they agree" do
+    context "with project, without tracking_id" do
       it "works" do
-        with_env("PROJECT" => "foo,bar", "TRACKING_ID" => "foo:x,bar:y") do
+        with_env("PROJECT" => "foo,bar", "TRACKING_ID" => nil) do
           f = Kennel::Filter.new
           f.project_filter.must_equal(["bar", "foo"])
-          f.tracking_id_filter.must_equal(["bar:y", "foo:x"])
+          f.tracking_id_filter.must_be_nil
         end
       end
     end
 
-    context "they disagree" do
-      with_env("PROJECT" => "foo,bar", "TRACKING_ID" => "foo:x,baz:y")
-
-      it "raises on .new" do
-        e = assert_raises(RuntimeError) do
+    context "with project, with tracking_id" do
+      it "works when they agree" do
+        f = with_env("PROJECT" => "foo,bar", "TRACKING_ID" => "foo:x,bar:y") do
           Kennel::Filter.new
         end
+        f.project_filter.must_equal(["bar", "foo"])
+        f.tracking_id_filter.must_equal(["bar:y", "foo:x"])
+      end
 
+      it "fails when they disagree" do
+        e = assert_raises(RuntimeError) do
+          with_env("PROJECT" => "foo,bar", "TRACKING_ID" => "foo:x,baz:y") do
+            Kennel::Filter.new
+          end
+        end
         e.message.must_include("do not set PROJECT= when using TRACKING_ID=")
       end
     end
+
+    context "without project, with tracking_id" do
+      it "works" do
+        with_env("PROJECT" => nil, "TRACKING_ID" => "foo:x,bar:y") do
+          f = Kennel::Filter.new
+          f.project_filter.must_equal(["bar", "foo"])
+          f.tracking_id_filter.must_equal(["bar:y", "foo:x"])
+        end
+      end
+    end
   end
 
-  context "without project, with tracking_id" do
-    it "works" do
-      with_env("PROJECT" => nil, "TRACKING_ID" => "foo:x,bar:y") do
-        f = Kennel::Filter.new
-        f.project_filter.must_equal(["bar", "foo"])
-        f.tracking_id_filter.must_equal(["bar:y", "foo:x"])
+  describe "#filter_projects!" do
+    it "filters nothing when not active" do
+      projects = [1]
+      Kennel::Filter.new.filter_projects!(projects)
+      projects.must_equal [1]
+    end
+
+    it "filters by project id" do
+      projects = [stub("P1", kennel_id: "a"), stub("P2", kennel_id: "b")]
+      with_env PROJECT: "a" do
+        Kennel::Filter.new.filter_projects!(projects)
+      end
+      projects.map(&:kennel_id).must_equal ["a"]
+    end
+  end
+
+  describe "#filter_parts!" do
+    it "filters nothing when not active" do
+      parts = [1]
+      Kennel::Filter.new.filter_parts!(parts)
+      parts.must_equal [1]
+    end
+
+    it "filters by project id" do
+      parts = [stub("P1", tracking_id: "a"), stub("P2", tracking_id: "b")]
+      with_env TRACKING_ID: "a" do
+        Kennel::Filter.new.filter_parts!(parts)
       end
+      parts.map(&:tracking_id).must_equal ["a"]
     end
   end
 
-  describe ".filter_resources!" do
+  describe "#filter_resources!" do
```

Comment:
Do we really want to be testing a private method? That can be OK, but is worth a comment to explain what's going on, and why we're doing that.

Reply:
- don't want to re-do the whole test
- it's kind of nice since the 2 public methods share it
... added a comment explaining

Reply:
Looks good.

This PR echoes work that I had already done, I remember writing the exact same code, and indeed even taking the same approach with testing via the private method – with a comment 😛 – but I didn't commit it / turn it into a PR because I chose to focus on refactoring the other areas first :-)